pkg/oc/cli/admin/mustgather/mustgather.go

```diff
@@ -0,0 +1,330 @@
+
+	return nil
+}
+
+type Options struct {
+	genericclioptions.IOStreams
+
+	Config *rest.Config
+	Client kubernetes.Interface
+
+	NodeName string
+	DestDir  string
+	Image    string
+	Command  []string
+	Keep     bool
+
+	RsyncRshCmd string
+}
+
+// Run creates and runs a must-gather pod.d
+func (o *Options) Run(rsyncCmd *cobra.Command) error {
+	if len(o.Config.BearerToken) != 0 {
+		return fmt.Errorf("cannot run must-gather with a token based kubeconfig")
+	}
+	if len(o.Image) == 0 {
+		return fmt.Errorf("missing an image")
+	}
+
+	var err error
+
+	// create namespace
+	ns, err := o.Client.CoreV1().Namespaces().Create(&corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "openshift-must-gather-",
+			Labels: map[string]string{
+				"openshift.io/run-level": "0",
+			},
+			Annotations: map[string]string{
+				"oc.openshift.io/command": "oc adm must-gather",
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if !o.Keep {
+		defer func() {
+			err = o.Client.CoreV1().Namespaces().Delete(ns.Name, nil)
+		}()
+	} else {
+		fmt.Fprintf(o.Out, "namespace/%s created.\n", ns.Name)
+	}
+
+	clusterRoleBinding, err := o.Client.RbacV1().ClusterRoleBindings().Create(o.newClusterRoleBinding(ns.Name))
+	if err != nil {
+		return err
+	}
+	if !o.Keep {
+		defer func() {
+			err = o.Client.RbacV1().ClusterRoleBindings().Delete(clusterRoleBinding.Name, &metav1.DeleteOptions{})
+		}()
+	} else {
+		fmt.Fprintf(o.Out, "clusterrolebinding.rbac.authorization.k8s.io/%s created.\n", clusterRoleBinding.Name)
+	}
+
+	// create pod
+	pod, err := o.Client.CoreV1().Pods(ns.Name).Create(o.newPod(o.NodeName))
+	if err != nil {
+		return err
+	}
+
+	// wait for pod to be running (gather has completed)
+	podPhase, err := o.waitForPodNotPending(pod)
+	if err != nil {
+		return err
+	}
+	if podPhase != corev1.PodRunning {
+		return fmt.Errorf("pod is not running: %v", podPhase)
+	}
+
+	// copy the gathered files into the local destination dir
+	err = o.copyFilesFromPod(pod)
+	return err
+}
+
+func (o *Options) copyFilesFromPod(pod *corev1.Pod) error {
+	rsyncOptions := &rsync.RsyncOptions{
+		Namespace:     pod.Namespace,
+		Source:        &rsync.PathSpec{PodName: pod.Name, Path: "/must-gather/"},
+		ContainerName: "copy",
+		Destination:   &rsync.PathSpec{PodName: "", Path: o.DestDir},
+		Client:        o.Client,
+		Config:        o.Config,
+		RshCmd:        fmt.Sprintf("%s --namespace=%s", o.RsyncRshCmd, pod.Namespace),
+		IOStreams:     o.IOStreams,
+	}
+	rsyncOptions.Strategy = rsync.NewDefaultCopyStrategy(rsyncOptions)
+	return rsyncOptions.RunRsync()
+
+}
+
+func (o *Options) waitForPodNotPending(pod *corev1.Pod) (corev1.PodPhase, error) {
+	watch, err := o.Client.CoreV1().Pods(pod.ObjectMeta.Namespace).Watch(metav1.ListOptions{
+		Watch:           true,
+		ResourceVersion: pod.ResourceVersion,
+		FieldSelector:   "metadata.name=" + pod.Name,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	phase := pod.Status.Phase
+	for {
+		select {
```

Comment:
Any particular reason we're not using `PollImmediate` here?

Reply:
Changed to `PollImmediate`.